### ISSUE TITLE
Strange interaction of @property and type parameters constrained to Never

### ISSUE THREAD

**Issue by lukas-lang** [microsoft/pyright#10950]
**Describe the bug**
I was trying to implement a class with members/properties that are "optionally Optional", which should be reflected in the type system. The attempt below uses the fact that `T | Never` is `T` to simply have a type parameter `N` that is either `None` (when the members should be of type `Optional[...]`) or `Never` (when they are not). As shown below, my initial attempt results in some very strange behavior in how the type of an `@property` member is resolved, consequently giving errors on assignments that are actually fine. I managed to remove the warnings by tweaking the code slightly (second snippet), but the behavior of the first still seemed worth reporting.

**Code or Screenshots**
The code ([Pyright Playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgA0UIpRAhgDYD68CRAsAFBdcDGDdAM6CoAMTBgA2gQBcACgJgURCsTIgAlAF0ZXKPqgAPGZhT4APoR6cDUAAIJwbPHD0GAJkWBQ4cwUQZgDSgAWgA%2BUwtCEwA6OLd9OzgY-xgYcgSoT29ff0CqEhNUKIINWPjuSuATcSlFZS0oAF4xCWklIi05DS4aMkYWRCI5YBjDHs4%2B%2BmZWYdG4CfnmqABGLiWW%2Bo4qgCYatrVyRpba6VoQLompgdmRnbGr2huhu5iF9fu4ZbXOYE-lrZcIA)):

```python
from typing import Never, reveal_type


class Foo[N:(None,Never)]:
    x: int | N

    @property
    def y(self) -> int | N: ...
    @y.setter
    def y(self, v: int | N): ...


f: Foo[None] = Foo[None]()
reveal_type(f.x)
reveal_type(f.y)
f.y = 1
f.y = None

f2: Foo[Never] = Foo[Never]()
reveal_type(f2.x)
reveal_type(f2.y)
f2.y = 1
f2.y = None
```

This shows a basic example of a class with a normal member of type `int | N` (with `N` either `None` or `Never`), and a `@property` of the same type. The code creates two instances for the two values of `N`, and tries to assign to the property. All but the last line (where I assign `None` to a property of type `int`) should pass without issue.

Pyright output:
> Type of "f.x" is "int | None"
Type of "f.y" is "int | None"
Type of "f2.x" is "int"
Type of "f2.y" is "int | N@Foo"
Cannot assign to attribute "y" for class "Foo[Never]"
  Argument of type "Foo[Never]" cannot be assigned to parameter "obj" of type "Foo[N@Foo]* | None" in function "__set__"
    Type "Foo[Never]" is not assignable to type "Foo[N@Foo]* | None"
      "Foo[Never]" is not assignable to "Foo[N@Foo]*"
        Type parameter "N@Foo" is invariant, but "Never" is not the same as "N@Foo"
      "Foo[Never]" is not assignable to "None"  (reportAttributeAccessIssue)
Cannot assign to attribute "y" for class "Foo[Never]"
  "None" is not assignable to "int"  (reportAttributeAccessIssue)

Notice how for `Foo[None]`, the revealed member/property types make sense (i.e. are both `int | None`), and the property assignment is also accepted.
For `Foo[Never]` however, the type of `f2.x` is `int` as expected, but `f2.y` is `int | N@Foo` for some reason (i.e. `N` was not properly substituted), and the property setting consequently fails. Interestingly, the error message for the last line is correct again, saying that `None` can't be assigned to `int` (so no mention of the unbound `N@Foo`)

The issue can be fixed by removing the explicit mention of `Never` in the constraint (since `Never` is the bottom type and thus a subtype of `None`, `Foo[Never]` is still accepted), giving the correct types & warnings ([Pyright Playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgA0UIpRAhgDYD68CRAsAFBdcDGDdAM6CoAMTBgA2gQBcBMCiIBdGVyjqoADxmYU%2BAD6EenDVAACCcGzxw1GgCZFgUOAApBRBsACUUALQAfLoGhDoAdBF26mZwYR4wMORRUI7Obh5eVCQ6qCEE3uGR3MXAOuJS8opKUAC8YhLSCsqu3lw0ZIwsiESuwGGarZzt9MysPX1wgxO1UACMXNN1lRwlAExlDcRkINV15dK0Oy1ttJ1jvav9g8Nn3RdhkwuXcDPznMDPM8tcQA)):

```python
from typing import Never, reveal_type


class Foo[N:None]:
    x: int | N

    @property
    def y(self) -> int | N: ...
    @y.setter
    def y(self, v: int | N): ...


f: Foo[None] = Foo[None]()
reveal_type(f.x)
reveal_type(f.y)
f.y = 1
f.y = None

f2: Foo[Never] = Foo[Never]()
reveal_type(f2.x)
reveal_type(f2.y)
f2.y = 1
f2.y = None
```

> 
Type of "f.x" is "int | None"
Type of "f.y" is "int | None"
Type of "f2.x" is "int"
Type of "f2.y" is "int"
Cannot assign to attribute "y" for class "Foo[Never]"
  "None" is not assignable to "int"  (reportAttributeAccessIssue)

**VS Code extension or command-line**
This is running on pyright-play.net, Pyright 1.1.405, so not related to any VS code specifics

**Comment by erictraut** [comment]
Value-constrained type variables are woefully underspecified in the typing spec, and they are misunderstood by most who attempt to use them. There are a number of fundamental problems with the concept of value-constrained type vars that I think will make it difficult or impossible to reconcile with the rest of the type system, so my general recommendation is to avoid using them altogether. It's almost always possible (as you've demonstrated above) to find a way to accomplish your goal without the use of constrained type variables. To my knowledge, no other language has this construct, which is good evidence that it's not needed.

A number tricky cases arise with value-constrained type vars when one of the constraints is a subtype of another constraint. This case (which shows up in your first example) should probably be considered an error case and disallowed by the typing spec. You will see inconsistent behaviors across type checkers in this case.

Your second example, which uses a type variable with an upper bound, looks like it's working correctly in compliance with the typing spec.

I'm going to close this issue because I don't plan to modify pyright's behavior for value-constrained type vars unless/until the typing spec is revised to clarify the intended behavior.
